sourced/ml/algorithms/uast2nodes.py

```diff
@@ -0,0 +1,131 @@
+from collections import defaultdict
+import numpy
+
+from sourced.ml.algorithms.uast_ids_to_bag import FakeVocabulary
+
+
+class UastNode2Bag():
+    """
+    Converts a UAST to a bag of features that are node specific.
+    The features can be either (internal types, number of children)
+    or (roles, number of children)
+    """
+    def __init__(self, type2ind=None, roles2ind=None, children2ind=None):
+        """
+        :param type2ind: The mapping from internal types to bag of keys. \
+            If None, no mapping is performed.
+        :param roles2ind: The mapping from roles to bag of keys.
+        :param children2ind: The mapping from the number of children to bag of keys. \
+            :class:'Quantization' is used if it is not specified.
+        """
+        self.type2ind = FakeVocabulary() if type2ind is None else type2ind
+        self.roles2ind = FakeVocabulary() if roles2ind is None else roles2ind
+        self.children2ind = Quantization() if children2ind is None else children2ind
+
+    def _uast2nodes(self, root):
+        """
+        :param uast: The UAST root node.
+        :return: The list of the nodes composing the UAST.
+        """
+        nodes = []
+        queue = [root]
+        while queue:
+            child = queue.pop(0)
+            queue.extend(child.children)
+            nodes.append(child)
+        return nodes
+
+    def _nodes2children(self, nodes):
+        """
+        :param nodes: List of all nodes from a UAST.
+        :return: The list of the number of children of all these nodes.
+        """
+        nb_children = []
+        for n in nodes:
+            nb_children.append(len(n.children))
+        return nb_children
+
+    def uast_to_bag(self, uast, feature):
+        """
+        Converts a UAST to a bag-of-features. The weights are feature frequencies.
+        The number of children are preprocessed by :class:`Quantization`
+
+        :param uast: The UAST root node.
+        :param feature: The first feature we want to extract from the nodes of the UAST.
+        :return:
+        """
+        bag = defaultdict(int)
+        nodes = self._uast2nodes(uast)
+        nb_children = self._nodes2children(nodes)
+        self.children2ind.build_partition(nb_children)
```

Comment:
We cannot do this at this stage: indices will be different for every file.

You need to build the global partition during the docfreq stage and assign it.